src/transformers/models/yoso/modeling_yoso.py

```diff
@@ -0,0 +1,1690 @@
+
+        self.use_expectation = config.use_expectation
+        self.hash_code_len = config.hash_code_len
+        self.use_conv = config.conv_window is not None
+        self.use_fast_hash = config.use_fast_hash
+        self.num_hash = config.num_hash
+        self.lsh_backward = config.lsh_backward
+
+        self.lsh_config = {
+            "hash_code_len": self.hash_code_len,
+            "use_fast_hash": self.use_fast_hash,
+            "num_hash": self.num_hash,
+            "lsh_backward": self.lsh_backward,
+        }
+
+        if config.conv_window is not None:
+            self.conv = nn.Conv2d(
+                in_channels=config.num_attention_heads,
+                out_channels=config.num_attention_heads,
+                kernel_size=(config.conv_window, 1),
+                padding=(config.conv_window // 2, 0),
+                bias=False,
+                groups=config.num_attention_heads,
+            )
+
+        if not self.use_expectation:
+            loaded = load_cuda_kernels()
+
+            if not loaded:
+                raise ValueError("""Failed to compile CUDA kernels. Ensure that the correct versions of PyTorch and cudatoolkit
+                    are installed and try again. Alternatively, avoid loading CUDA kernels by using YOSO Expectation by
+                    setting use_expectation=True.""")
+
+    def transpose_for_scores(self, layer):
+        new_layer_shape = layer.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        layer = layer.view(*new_layer_shape)
+        return layer.permute(0, 2, 1, 3)
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_value=None,
+        output_attentions=False,
+    ):
+        mixed_query_layer = self.query(hidden_states)
+
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
+        value_layer = self.transpose_for_scores(self.value(hidden_states))
+        query_layer = self.transpose_for_scores(mixed_query_layer)
+
+        if self.use_conv:
+            conv_value_layer = self.conv(value_layer * attention_mask[:, None, :, None])
+
+        batch_size, num_heads, seq_len, head_dim = query_layer.size()
+
+        query_layer = query_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+        key_layer = key_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+        value_layer = value_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+
+        # revert changes made by get_extended_attention_mask
+        attention_mask = 1.0 + attention_mask / 10000.0
+        attention_mask = attention_mask.squeeze().repeat(1, num_heads, 1).reshape(batch_size * num_heads, seq_len)
+
+        if (not self.use_expectation) and head_dim < 32:
```

Comment:
(nit) the 32 is a bit random here to me - why 32? Could we maybe give it a variable name?

Reply:
and also replace it with the 32's below

Reply:
The CUDA kernels are implemented such that they work best with inputs whose size is a multiple of a GPU's warp size - 32. Inputs smaller than this are padded to 32. I included a comment about this in the code.